app/services/fact_importer.rb

```diff
@@ -91,16 +93,18 @@ def create_fact_name(fact_names, name, fact_value)
   end
 
   def update_facts
-    facts_to_update = []
-    db_facts.each { |name, fv| facts_to_update << [facts[name], fv] if fv.value != facts[name] }
-
-    @counters[:updated] = facts_to_update.size
-    return logger.debug("No facts update required for #{host}") if facts_to_update.empty?
-
-    logger.debug("Merging facts for '#{host}': updated #{@counters[:updated]} facts")
-
-    facts_to_update.each do |new_value, fv|
-      fv.update_attribute(:value, new_value)
+    @counters[:updated] = 0
+    db_facts_as_relation.find_each do |record|
+      new_value = facts[record.name]
+      if record.value != new_value
+        record.update_attribute(:value, new_value)
```

Comment:
hence why i'm asking for the difference, I'm not sure we actually have any
callbacks (besides people who might use foreman hooks) but avoiding it does
not let us handle events in the future (e.g. if we want a notification of
certain fact changes etc).